### PR TITLE
8278171: [vectorapi] Mask incorrectly computed for zero extending cast

### PR DESCRIPTION
**Pull request by merykitty**
When doing an unsigned upcast, we perform a signed cast followed by a bitwise and operation to clip the extended signed bit. The sign clip mask is currently calculated incorrectly, the correct mask should have the number of least significant bit set equal to the size of the source elements. This patch fixes this trivial issue and adds several tests for zero extension casts.

Thank you very much.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8278171](https://bugs.openjdk.java.net/browse/JDK-8278171): [vectorapi] Mask incorrectly computed for zero extending cast

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6634/head:pull/6634` \
`$ git checkout pull/6634`

Update a local copy of the PR: \
`$ git checkout pull/6634` \
`$ git pull https://git.openjdk.java.net/jdk pull/6634/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6634`

View PR using the GUI difftool: \
`$ git pr show -t 6634`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6634.diff">https://git.openjdk.java.net/jdk/pull/6634.diff</a>

</details>

### CODE REVIEW

**Comment by merykitty** [comment]
@PaulSandoz Could you take a look at this PR? Also, could you create an issue for this PR, please.
Should this be split into 2, the first one fixes the bug and add tests while the second one implements the intrinsics.
Thank you very much.

File: src/hotspot/cpu/x86/x86.ad [hunk 1]
```diff
@@ -1799,6 +1802,21 @@ const bool Matcher::match_rule_supported_vector(int opcode, int vlen, BasicType
...
+    case Op_VectorUCastS2X:
+      if (size_in_bits == 256 && UseAVX < 2) {
+        return false;
+      }
+      break;
+    case Op_VectorUCastI2X:
+      if (size_in_bits == 256 && UseAVX < 2) {
+        return false;
+      }
+      break;
```
Comment (PaulSandoz) [inline src/hotspot/cpu/x86/x86.ad:1819]: Collapse cases, since each has the code code?

**Review by PaulSandoz — COMMENTED** [review]
I am inclined to separated out. Fix the bug, add the tests, and integrate for 18. Then enhance with the intrinsics for 19.

If you agree to that I will create two bugs.

**Comment by merykitty** [comment]
@PaulSandoz Yes, I think that should be the case, thank you very much.

**Comment by PaulSandoz** [comment]
@merykitty here you go:

[vectorapi] Mask incorrectly computed for zero extending cast
https://bugs.openjdk.java.net/browse/JDK-8278171

[vectorapi] Add x64 intrinsics for unsigned (zero extended) casts
https://bugs.openjdk.java.net/browse/JDK-8278173

**Comment by PaulSandoz** [comment]
Looks good. Running tests. 

Something we should consider later: split this test out into smaller files and support the max vector. On supported platforms we could consider using the IR test framework to verify the correct IR node is generated.

**Review by PaulSandoz — APPROVED** [review]
Tests are good.

**Comment by merykitty** [comment]
Thanks a lot for your support and review. Do I need another review here?

**Comment by PaulSandoz** [comment]
> Thanks a lot for your support and review. Do I need another review here?

No need for this one, for next PR for the intrinsics will likely need two HotSpot reviewers.

**Pull request closed without merging**